File: app/src/main/java/com/duzhaokun123/bilibilihd/ui/userSpace/UserSpaceActivity.java
package com.duzhaokun123.bilibilihd.ui.userSpace;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import android.content.Intent;
import android.os.Bundle;
import android.os.Message;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.duzhaokun123.bilibilihd.R;
import com.duzhaokun123.bilibilihd.myBilibiliApi.space.SpaceAPI;
import com.duzhaokun123.bilibilihd.myBilibiliApi.space.model.Space;
import com.duzhaokun123.bilibilihd.pBilibiliApi.api.PBilibiliClient;
import com.duzhaokun123.bilibilihd.ui.PhotoViewActivity;
import com.duzhaokun123.bilibilihd.utils.ToastUtil;
import com.google.android.material.tabs.TabLayout;
//import com.hiczp.bilibili.api.app.model.Space;

import de.hdodenhof.circleimageview.CircleImageView;

import static com.duzhaokun123.bilibilihd.utils.OtherUtils.setLevelDrawable;
import static com.duzhaokun123.bilibilihd.utils.OtherUtils.setSixDrawable;

public class UserSpaceActivity extends AppCompatActivity {

    private ImageView mIvSpaceImage, mIvSex,mIvLevel;
    private CircleImageView mCivFace;
    private TextView mTvName, mTvFans, mTvWatching, mTvLike, mTvSign;
    private TabLayout mTl;
    private ViewPager mVp;

    private Bundle bundle;
    private PBilibiliClient pBilibiliClient;
    private Space mSpace;
    private Handler handler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_space);

        mIvSpaceImage = findViewById(R.id.iv_spaceImage);
        mIvSex = findViewById(R.id.iv_sex);
        mIvLevel = findViewById(R.id.iv_level);
        mCivFace = findViewById(R.id.civ_face);
        mTvName = findViewById(R.id.tv_name);
        mTvFans = findViewById(R.id.tv_fans);
        mTvWatching = findViewById(R.id.tv_watching);
        mTvLike = findViewById(R.id.tv_like);
        mTvSign = findViewById(R.id.tv_sign);
        mTl = findViewById(R.id.tl);
        mVp = findViewById(R.id.vp);

        bundle = getIntent().getExtras();

        handler = new Handler();

        pBilibiliClient = PBilibiliClient.Companion.getPBilibiliClient();
        new Thread(new Runnable() {
            @Override
            public void run() {
//                try{
//                    mSpace = pBilibiliClient.getPAppAPI().space(bundle.getLong("uid"));
//                    handler.sendEmptyMessage(0);
//                } catch (Exception e) {
//                    e.printStackTrace();
//                    Looper.prepare();
//                    ToastUtil.sendMsg(UserSpaceActivity.this, e.getMessage());
//                    Looper.loop();
//
//                }
                SpaceAPI spaceAPI = SpaceAPI.getSpaceAPI();
                spaceAPI.getSpace(bundle.getLong("uid"), new SpaceAPI.Callback() {
                    @Override
                    public void onException(Exception e) {
                        e.printStackTrace();
                        ToastUtil.sendMsg(UserSpaceActivity.this, e.getMessage());
                    }

                    @Override
                    public void onSuccess(com.duzhaokun123.bilibilihd.myBilibiliApi.space.model.Space space) {
                        mSpace = space;
                        handler.sendEmptyMessage(0);
                    }
                });
            }
        }).start();

        mVp.setAdapter(new MyFragmentPagerAdapter(getSupportFragmentManager(), FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT));
        mTl.setupWithViewPager(mVp);

        mCivFace.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(UserSpaceActivity.this, PhotoViewActivity.class);
                intent.putExtra("url", mSpace.getData().getCard().getFace());
                startActivity(intent);
            }
        });
    }

    class MyFragmentPagerAdapter extends FragmentPagerAdapter {

        private Fragment mHomeFragment, mTrendFragment, mSubmitFragment, mFavoriteFragment;

        public MyFragmentPagerAdapter(@NonNull FragmentManager fm, int behavior) {
            super(fm, behavior);
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    if (mHomeFragment == null) {
                        mHomeFragment = new HomeFragment(mSpace);
                    }
                    return mHomeFragment;
                case 1:
                    if (mTrendFragment == null) {
                        mTrendFragment = new TrendFragment(mSpace);
                    }
                    return mTrendFragment;
                case 2:
                    if (mSubmitFragment == null) {
                        mSubmitFragment = new SubmitFragment(mSpace);
                    }
                    return mSubmitFragment;
                case 3:
                    if (mFavoriteFragment == null) {
                        mFavoriteFragment = new FavoriteFragment(mSpace);
                    }
                    return mFavoriteFragment;
            }

            return null;
        }

        @Override
        public int getCount() {
            return 4;
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return getString(R.string.home);
                case 1:
                    return getString(R.string.dynamic);
                case 2:
                    return getString(R.string.submit);
                case 3:
                    return getString(R.string.favorite);
            }
            return super.getPageTitle(position);
        }
    }

    class Handler extends android.os.Handler{
        @Override
        public void handleMessage(@NonNull Message msg) {
            Glide.with(mCivFace).load(mSpace.getData().getCard().getFace()).into(mCivFace);
            if (mSpace.getData().getImages().getImgUrl().equals("")) {
                Glide.with(mIvSpaceImage).load("https://i0.hdslb.com/bfs/space/cb1c3ef50e22b6096fde67febe863494caefebad.png").into(mIvSpaceImage);
                mIvSpaceImage.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Intent intent = new Intent(UserSpaceActivity.this, PhotoViewActivity.class);
                        intent.putExtra("url", "https://i0.hdslb.com/bfs/space/cb1c3ef50e22b6096fde67febe863494caefebad.png");
                        startActivity(intent);
                    }
                });
            } else {
                Glide.with(mIvSpaceImage).load(mSpace.getData().getImages().getImgUrl()).into(mIvSpaceImage);
                mIvSpaceImage.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Intent intent = new Intent(UserSpaceActivity.this, PhotoViewActivity.class);
                        intent.putExtra("url", mSpace.getData().getImages().getImgUrl());
                        startActivity(intent);
                    }
                });
            }
            setSixDrawable(mIvSex, mSpace.getData().getCard().getSex());
            setLevelDrawable(mIvLevel, mSpace.getData().getCard().getLevelInfo().getCurrentLevel());
            mTvName.setText(mSpace.getData().getCard().getName());
            mTvFans.setText(String.valueOf(mSpace.getData().getCard().getFans()));
            mTvSign.setText(mSpace.getData().getCard().getSign());
            mTvWatching.setText(String.valueOf(mSpace.getData().getCard().getAttention()));
            if (mSpace.getData().getCard().getVip().getVipType() != 1) {
                mTvName.setTextColor(getColor(R.color.colorAccent));
            }
            // TODO: 20-2-22 试出 mTvLike 对应的键
        }
    }
}

File: app/src/main/java/com/duzhaokun123/bilibilihd/ui/userspace/UserSpaceActivity.kt
package com.duzhaokun123.bilibilihd.ui.userspace

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.graphics.drawable.Drawable
import android.util.Pair
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.TextView
import androidx.core.view.updatePadding
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.bumptech.glide.Glide
import com.duzhaokun123.bilibilihd.Application
import com.duzhaokun123.bilibilihd.R
import com.duzhaokun123.bilibilihd.bases.BaseActivity2
import com.duzhaokun123.bilibilihd.databinding.ActivityUserSpaceBinding
import com.duzhaokun123.bilibilihd.utils.*
import com.duzhaokun123.bilibilihd.utils.ImageViewUtil.setLevelDrawable
import com.duzhaokun123.bilibilihd.utils.ImageViewUtil.setSixDrawable
import com.duzhaokun123.bilibilihd.utils.ImageViewUtil.viewImage
import com.duzhaokun123.bilibilihd.utils.LinkifyUtil.addAllLinks
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.hiczp.bilibili.api.app.model.Space
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*

class UserSpaceActivity : BaseActivity2<ActivityUserSpaceBinding>() {
    private var mSpace: Space? = null
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.user_space_activity, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.open_in_browser) {
            BrowserUtil.openCustomTab(this, MyBilibiliClientUtil.getUserSpaceLink(startIntent.getLongExtra(EXTRA_UID, 0)))
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun initConfig() = setOf(Config.TRANSPARENT_ACTION_BAR)

    override fun initLayout() = R.layout.activity_user_space

    @SuppressLint("SetTextI18n")
    override fun initView() {
        if (Settings.layout.isUserSpaceUseWebView) {
            BrowserUtil.openWebViewActivity(this, "https://space.bilibili.com/${startIntent.getLongExtra(EXTRA_UID, 0)}", true, true)
            finish()
            return
        }
        baseBind.civFace.setImageDrawable(ObjectCache.get(startIntent.getStringExtra(EXTRA_FACE_CACHE_ID)) as Drawable?)
        baseBind.tvName.text = ObjectCache.get(startIntent.getStringExtra(EXTRA_NAME_CACHE_ID)) as CharSequence?
        baseBind.tvUid.text = "UID: " + startIntent.getLongExtra(EXTRA_UID, 0)
        baseBind.vp.adapter = MyFragmentStateAdapter(this)
        TabLayoutMediator(baseBind.tl, baseBind.vp) { tab: TabLayout.Tab, position: Int ->
            when (position) {
                0 -> tab.setText(R.string.home)
                1 -> tab.setText(R.string.dynamic)
                2 -> tab.setText(R.string.submit)
                3 -> tab.setText(R.string.favorite)
            }
        }.attach()
        baseBind.civFace.setOnClickListener {
            if (mSpace != null) {
                viewImage(this, mSpace!!.data.card.face, baseBind.civFace, true)
            }
        }
        title = null
        baseBind.tvFans.setOnClickListener {
            BrowserUtil.openWebViewActivity(this,
                    "https://space.bilibili.com/h5/follow?type=fans&mid=${startIntent.getLongExtra(EXTRA_UID, 0)}&night=${if (OtherUtils.isNightMode()) 1 else 0}", false, true)
        }
        baseBind.tvWatching.setOnClickListener {
            BrowserUtil.openWebViewActivity(this,
                    "https://space.bilibili.com/h5/follow?type=follow&mid=${startIntent.getLongExtra(EXTRA_UID, 0)}&night=${if (OtherUtils.isNightMode()) 1 else 0}", false, true)
        }
    }

    override fun initData() {
        if (Settings.layout.isUserSpaceUseWebView) return

        GlobalScope.launch(Dispatchers.IO) {
            try {
                mSpace = Application.getPBilibiliClient().pAppAPI.space(startIntent.getLongExtra(EXTRA_UID, 0))

            } catch (e: Exception) {
                e.printStackTrace()
                kRunOnUiThread { TipUtil.showToast(e.message) }
            }
            if (mSpace != null) {
                kRunOnUiThread { setInfo() }
            }
        }
    }

    private fun setInfo() {
        Glide.with(this).load(mSpace!!.data.card.face).into(baseBind.civFace)
        if (mSpace!!.data.images.imgUrl == "") {
            Glide.with(this).load("https://i0.hdslb.com/bfs/space/cb1c3ef50e22b6096fde67febe863494caefebad.png").into(baseBind.ivSpaceImage)
            baseBind.ivSpaceImage.setOnClickListener { viewImage(this, "https://i0.hdslb.com/bfs/space/cb1c3ef50e22b6096fde67febe863494caefebad.png", baseBind.ivSpaceImage, true) }
        } else {
            Glide.with(this).load(mSpace!!.data.images.imgUrl).into(baseBind.ivSpaceImage)
            baseBind.ivSpaceImage.setOnClickListener { viewImage(this, mSpace!!.data.images.imgUrl, baseBind.ivSpaceImage, true) }
        }
        baseBind.space = mSpace
        setSixDrawable(baseBind.ivSex, mSpace!!.data.card.sex)
        setLevelDrawable(baseBind.ivLevel, mSpace!!.data.card.levelInfo.currentLevel)
        addAllLinks(baseBind.tvSign)
        if (mSpace!!.data.card.vip.vipType != 0) {
            baseBind.tvName.setTextColor(getColor(R.color.colorAccent))
        }
    }

    internal inner class MyFragmentStateAdapter(fragmentActivity: FragmentActivity) : FragmentStateAdapter(fragmentActivity) {
        private val mHomeFragment by lazy { HomeFragment(mSpace) }
        private val mTrendFragment by lazy { DynamicFragment(mSpace) }
        private val mSubmitFragment by lazy { SubmitFragment(mSpace) }
        private val mFavoriteFragment by lazy { FavoriteFragment(mSpace) }

        override fun createFragment(position: Int): Fragment {
            return when (position) {
                0 -> mHomeFragment
                1 -> mTrendFragment
                2 -> mSubmitFragment
                3 -> mFavoriteFragment
                else -> Fragment()
            }
        }

        override fun getItemCount() = 4
    }

    companion object {
        const val EXTRA_UID = "uid"
        const val EXTRA_FACE_CACHE_ID = "face_cache_id"
        const val EXTRA_NAME_CACHE_ID = "name_cache_id"

        private fun enter(activity: Activity, uid: Long,
                          faceView: View?, faceDrawable: Drawable?,
                          nameView: View?, nameCS: CharSequence?) {
            val intent = Intent(activity, UserSpaceActivity::class.java)
            intent.putExtra(EXTRA_UID, uid)
            if (Settings.layout.isUserSpaceUseWebView) {
                BrowserUtil.openWebViewActivity(activity, "https://space.bilibili.com/$uid", true, true)
            } else {
                intent.putExtra(EXTRA_FACE_CACHE_ID, ObjectCache.put(faceDrawable))
                intent.putExtra(EXTRA_NAME_CACHE_ID, ObjectCache.put(nameCS))
                val pairs = ArrayList<Pair<View, String>>()
                if (faceView != null) {
                    pairs.add(Pair.create(faceView, "face"))
                }
                if (nameView != null) {
                    pairs.add(Pair.create(nameView, "name"))
                }
                activity.startActivity(intent, ActivityOptions.makeSceneTransitionAnimation(activity, *pairs.toTypedArray()).toBundle())
            }
        }

        @JvmStatic
        @JvmOverloads
        fun enter(activity: Activity, uid: Long, faceView: ImageView? = null, nameView: TextView? = null) {
            enter(activity, uid, faceView, faceView?.drawable, nameView, nameView?.text)
        }
    }
}